File: lib/libesp32/Berry/tools/coc/block_builder.cpp
/********************************************************************
** Copyright (c) 2018-2020 Guan Wenliang
** This file is part of the Berry default interpreter.
** skiars@qq.com, https://github.com/Skiars/berry
** See Copyright Notice in the LICENSE file or at
** https://github.com/Skiars/berry/blob/master/LICENSE
********************************************************************/
#include "block_builder.h"
#include "hash_map.h"
#include "macro_table.h"
#include "object_block.h"
#include <sstream>
#include <fstream>

static bool depend(const object_block *object, const macro_table *macro)
{
    auto it = object->attr.find("depend");
    if (it != object->attr.end()) {
        return macro->query(it->second);
    }
    return true;
}

block_builder::block_builder(const object_block *object, const macro_table *macro)
{
    m_block.name = object->name;
    if (depend(object, macro)) {
        m_block.type = object->type;
        m_block.attr = object->attr;

        auto it = object->attr.find("name");
        if (it != object->attr.end()) {
            m_strtab.push_back(it->second);
        }
        
        for (auto i : object->data) {
            if (i.second.depend.empty() || macro->query(i.second.depend)) {
                m_block.data[i.first] = i.second.value;
                m_strtab.push_back(i.first);
            }
        }
    }
}

std::string block_builder::block_tostring(const block &block)
{
    std::ostringstream ostr;
    if (block.type == "map") {
        ostr << map_tostring(block, block.name);
    } else if (block.type == "class") {
        ostr << class_tostring(block);
    } else if (block.type == "vartab") {
        ostr << vartab_tostring(block);
    } else if (block.type == "module") {
        ostr << module_tostring(block);
    }
    return ostr.str();
}

std::string block_builder::class_tostring(const block &block)
{
    bool empty_map = block.data.empty();
    std::ostringstream ostr;
    hash_map map(block.data);
    std::string map_name(block.name + "_map");

    if (!empty_map) {
        ostr << map_tostring(block, map_name, true) << std::endl;
    }
    ostr << scope(block) << " be_define_const_class(\n    "
         << block.name << ",\n    "
         << map.var_count() << ",\n    "
         << super(block) << ",\n    "
         << name(block) << "\n"
            ");" << std::endl;
    return ostr.str();
}

std::string block_builder::map_tostring(const block &block, const std::string &name, bool local)
{
    std::ostringstream ostr;
    hash_map map(block.data);

    hash_map::entry_table list = map.entry_list();
    ostr << "static be_define_const_map_slots(" << name << ") {\n";
    for (auto it : list) {
        ostr << "    { be_const_key(" << it.key << ", "
             << it.next << "), " << it.value << " }," << std::endl;
    }
    ostr << "};\n\n";

    ostr << (local ? "static" : scope(block))
         << " be_define_const_map(\n    "
         << name << ",\n    "
         << list.size() << "\n"
            ");" << std::endl;
    return ostr.str();
}

std::string block_builder::vartab_tostring(const block &block)
{
    std::ostringstream ostr;
    struct block idxblk;
    std::vector<std::string> varvec;
    int index = 0;

    idxblk = block;
    idxblk.data.clear();
    for (auto it : block.data) {
        varvec.push_back(it.second);
        it.second = "int(" + std::to_string(index++) + ")";
        idxblk.data.insert(it);
    }

    ostr << map_tostring(idxblk, block.name + "_map", true) << std::endl;
    ostr << "static const bvalue __vlist_array[] = {\n";
    for (auto it : varvec) {
        ostr << "    be_const_" << it << "," << std::endl;
    }
    ostr << "};\n\n";

    ostr << "static be_define_const_vector(\n    "
         << block.name << "_vector,\n    "
            "__vlist_array,\n    "
         << varvec.size() << "\n"
            ");" << std::endl;
    return ostr.str();
}

std::string block_builder::module_tostring(const block &block)
{
    std::ostringstream ostr;
    std::string name("m_lib" + block.name);
    std::string map_name(name + "_map");

    ostr << map_tostring(block, map_name, true) << std::endl
         << "static be_define_const_module(\n    "
         << name << ",\n    "
            "\"" << block.name << "\"\n"
            ");" << std::endl;
    std::string scp = scope(block);
    if (scp != "static") { /* extern */
        ostr << "\n" << scp
             << " be_define_const_native_module("
             << block.name << ", "
             << init(block) << ");" << std::endl;
    }
    return ostr.str();
}

std::string block_builder::scope(const block &block)
{
    auto it = block.attr.find("scope");
    return it != block.attr.end() && it->second == "local" ?
        "static" : "BE_EXPORT_VARIABLE";
}

std::string block_builder::super(const block &block)
{
    auto it = block.attr.find("super");
    return it == block.attr.end() ? "NULL" : "(bclass *)&" + it->second;
}

std::string block_builder::name(const block &block)
{
    auto it = block.attr.find("name");
    return it == block.attr.end() ? block.name : it->second;
}

std::string block_builder::init(const block &block)
{
    auto it = block.attr.find("init");
    return it == block.attr.end() ? "NULL" : it->second;
}

void block_builder::writefile(const std::string &filename, const std::string &text)
{
    std::string pathname(filename);
    std::string otext("#include \"be_constobj.h\"\n\n" + text);

    std::ostringstream buf;
    std::ifstream fin(pathname);
    buf << fin.rdbuf();
    if (buf.str() != otext) {
        std::ofstream fout;
        fout.open(pathname, std::ios::out);
        fout << otext;
        fout.close();
    }
}

void block_builder::dumpfile(const std::string &path)
{
    std::string s = block_tostring(m_block);
    auto it = m_block.attr.find("file");
    std::string &name = it != m_block.attr.end() ? it->second : m_block.name;
    writefile(path + "/be_fixed_" + name + ".h", s);
}

const std::vector<std::string>& block_builder::strtab() const
{
    return m_strtab;
}

File: lib/libesp32/berry/tools/pycoc/block_builder.py
import copy
from hash_map import *

class block:
    def __init__(self):
        self.type = ""
        self.name = ""
        self.attr = {}
        self.data = {}
        self.data_ordered = []

def depend(obj, macro):
    if "depend" in obj.attr:
        return macro.query(obj.attr["depend"])
    else:
        return True

class block_builder:
    """Output an object"""

    def __init__(self, obj, macro):
        self.block = block()
        self.strtab = []

        self.block.name = obj.name
        if depend(obj, macro):
            self.block.type = obj.type
            self.block.attr = obj.attr
            
            if "name" in obj.attr:
                self.strtab.append(obj.attr["name"])
            
            for key in obj.data_ordered:
                second = obj.data[key]
                if second.depend == None or macro.query(second.depend):
                    self.block.data[key] = second.value
                    self.strtab.append(key)
                    self.block.data_ordered.append(key)
    
    def block_tostring(self, block):
        ostr = ""
        if block.type == "map":
            ostr += self.map_tostring(block, block.name)
        elif block.type == "class":
            ostr += self.class_tostring(block)
        elif block.type == "vartab":
            ostr += self.vartab_tostring(block)
        elif block.type == "module":
            ostr += self.module_tostring(block)
        return ostr

    def class_tostring(self, block):
        ostr = ""
        hmap = hash_map(block.data)
        map_name = block.name + "_map"
        if len(block.data) > 0:
            ostr += self.map_tostring(block, map_name, True) + "\n"
        
        ostr += self.scope(block) + " be_define_const_class(\n    "
        ostr += block.name + ",\n    "
        ostr += str(hmap.var_count()) + ",\n    "
        ostr += self.get_super(block) + ",\n    "
        ostr += self.name(block) + "\n);\n"
        return ostr
    
    def map_tostring(self, block, name, local):
        hmap = hash_map(block.data)
        entlist = hmap.entry_list()
        ostr = "static be_define_const_map_slots(" + name + ") {\n"
        for ent in entlist:
            ostr += "    { be_const_key(" + ent.key + ", "
            ostr += str(ent.next) + "), " + ent.value + " },\n"
        ostr += "};\n\n"

        if local:
            ostr += "static"
        else:
            ostr += self.scope(block)
        ostr += " be_define_const_map(\n    "
        ostr += name + ",\n    "
        ostr += str(len(entlist)) + "\n);\n"
        return ostr
    
    def vartab_tostring(self, block):
        ostr = ""
        varvec = []
        idxblk = copy.deepcopy(block)
        index = 0
        idxblk.data = {}
        for key in block.data_ordered:
            varvec.append(block.data[key])
            idxblk.data[key] = "int(" + str(index) + ")"
            index += 1
        
        ostr += self.map_tostring(idxblk, block.name + "_map", True) + "\n"
        ostr += "static const bvalue __vlist_array[] = {\n";
        for it in varvec:
            ostr += "    be_const_" + it + ",\n"
        ostr += "};\n\n"

        ostr += "static be_define_const_vector(\n    "
        ostr += block.name + "_vector,\n    __vlist_array,\n    "
        ostr += str(len(varvec)) + "\n);\n"
        return ostr
    
    def module_tostring(self, block):
        ostr = ""
        name = "m_lib" + block.name
        map_name = name + "_map"

        ostr += self.map_tostring(block, map_name, True) + "\n"
        ostr += "static be_define_const_module(\n    "
        ostr += name + ",\n    "
        ostr += "\"" + block.name + "\"\n);\n"
        scp = self.scope(block)
        if scp != "static":
            ostr += "\n" + scp
            ostr += " be_define_const_native_module("
            ostr += block.name + ");\n"
        return ostr

    def scope(self, block):
        if "local" in block.attr:
            return "static"
        else:
            return "BE_EXPORT_VARIABLE"

    def get_super(self, block):
        if "super" in block.attr:
            return "(bclass *)&" + block.attr["super"]
        else:
            return "NULL"
    
    def name(self, block):
        if "name" in block.attr:
            return block.attr["name"]
        else:
            return block.name
    
    def writefile(self, filename, text):
        otext = "#include \"be_constobj.h\"\n\n" + text
        with open(filename, "w") as f:
            f.write(otext)

    def dumpfile(self, path):
        s = self.block_tostring(self.block)
        if "file" in self.block.attr:
            name = self.block.attr["file"]
        else:
            name = self.block.name
        self.writefile(path + "/be_fixed_" + name + ".h", s)
        